libraries/rush-lib/src/logic/PackageJsonUpdater.ts

```diff
@@ -151,6 +181,158 @@ export class PackageJsonUpdater {
+
+      const version: string = await this._getNormalizedVersionSpec(
+        projects,
+        moduleName,
+        latestVersion,
+        implicitlyPreferredVersion,
+        explicitlyPreferredVersion,
+        inferredRangeStyle
+      );
+
+      if (devDependency) {
+        devDependenciesToUpdate[moduleName] = version;
+      } else {
+        dependenciesToUpdate[moduleName] = version;
+      }
+
+      this._terminal.writeLine(
+        colors.green(`Updating projects to use `) + moduleName + '@' + colors.cyan(version)
+      );
+      this._terminal.writeLine();
+
+      const existingSpecifiedVersions: Set<string> | undefined = allVersionsByPackageName.get(moduleName);
+      if (
+        existingSpecifiedVersions &&
+        !existingSpecifiedVersions.has(version) &&
+        this._rushConfiguration.ensureConsistentVersions &&
+        !updateOtherPackages
+      ) {
+        // There are existing versions, and the version we're going to use is not one of them, and this repo
+        // requires consistent versions, and we aren't going to update other packages, so we can't proceed.
+
+        const existingVersionList: string = Array.from(existingSpecifiedVersions).join(', ');
+        throw new Error(
+          `Adding '${moduleName}@${version}' ` +
+            `causes mismatched dependencies. Use the "--make-consistent" flag to update other packages to use ` +
+            `this version, or try specify one of the existing versions (${existingVersionList}).`
+        );
+      }
+    }
+
+    const allPackageUpdates: Map<string, VersionMismatchFinderEntity> = new Map();
+    const allDependenciesToUpdate: [string, string][] = [
+      ...Object.entries(dependenciesToUpdate),
+      ...Object.entries(devDependenciesToUpdate)
+    ];
+
+    for (const project of projects) {
+      const mismatchFinderProject: VersionMismatchFinderProject = new VersionMismatchFinderProject(project);
+
+      const currentProjectDepUpdate: IUpdateProjectOptions = {
+        project: mismatchFinderProject,
+        dependenciesToAddOrUpdateOrRemove: dependenciesToUpdate,
+        dependencyType: DependencyType.Regular
+      };
+
+      const currentProjectDevDepUpdate: IUpdateProjectOptions = {
+        project: mismatchFinderProject,
+        dependenciesToAddOrUpdateOrRemove: devDependenciesToUpdate,
+        dependencyType: DependencyType.Dev
+      };
+
+      allPackageUpdates.set(mismatchFinderProject.filePath, mismatchFinderProject);
+
+      this.updateProject(currentProjectDepUpdate);
+      this.updateProject(currentProjectDevDepUpdate);
+    }
+
+    if (this._rushConfiguration.ensureConsistentVersions || updateOtherPackages) {
+      // TODO: Reenable to support mismatch checks for both devDeps and regularDeps
+      // we need to do a mismatch check
+      const mismatchFinder: VersionMismatchFinder = VersionMismatchFinder.getMismatches(
+        this._rushConfiguration,
+        {
+          variant: variant
+        }
+      );
```

Comment:
Move this into the `if` below.
At which point, there's really only one `if`, since the outer one is redundant.

Reply:
We need to check both cases from the outer `if` so just dropping the inner `if` should be fine right?

Reply:
Will you recheck change in latest commit before I resolve to verify.

Reply:
The outer if is the one that can be dropped. Since the `getMismatches()` call doesn't have side effects, it can be moved into the inner `if`.

Reply:
Done and done. Removed outer `if`, moved getMismatches to inner `if` at both callsites.